### PR TITLE
#36: Split the functionality to HasValues and HasValuesMatching

### PR DESCRIPTION
**Pull request by dgroup**
#36

### CODE REVIEW

**Comment by 0crat** [comment]
Job [#43](https://github.com/llorllale/cactoos-matchers/issues/43) is now in scope, role is `REV`

File: src/main/java/org/llorllale/cactoos/matchers/HasValues.java [hunk 1]
```diff
@@ -26,72 +26,72 @@
...
  * @checkstyle ProtectedMethodInFinalClassCheck (200 lines)
  */
 public final class HasValues<X> extends TypeSafeDiagnosingMatcher<Iterable<X>> {
+
     /**
-     * The function to check the {@link Iterable}.
+     * The expected values within the collection.
      */
-    private final UncheckedFunc<X, Boolean> fnc;
+    private final Collection<X> expected;
```
Comment (llorllale) [inline src/main/java/org/llorllale/cactoos/matchers/HasValues.java]: @dgroup we frown upon the use of `Collection`, preferring `Iterable` instead

File: src/main/java/org/llorllale/cactoos/matchers/HasValues.java [hunk 2]
```diff
@@ -26,72 +26,72 @@
...
-    protected boolean matchesSafely(final Iterable<X> actual,
+    protected boolean matchesSafely(final Iterable<X> item,
         final Description dsc) {
-        dsc.appendText("The function applied to ")
-            .appendValue(new ListOf<>(actual))
-            .appendText(" is failed");
-        return new UncheckedScalar<>(
-            new Or(this.fnc, actual)
-        ).value();
+        final Collection<X> actual = new CollectionOf<>(item);
```
Comment (llorllale) [inline src/main/java/org/llorllale/cactoos/matchers/HasValues.java]: @dgroup this code can be made elegant by using some of cactoos primitives... think you can do that?

File: src/test/java/org/llorllale/cactoos/matchers/HasValuesTest.java [hunk 3]
```diff
@@ -42,81 +40,74 @@
...
-     * @todo #32:30m Matcher to test {@link org.hamcrest.Matcher} objects.
-     *  In order to check the matchers we should verify that
-     *  - error message is correct in case false status
-     *  - the status of testing is correct itself.
-     *  The possible name - {@code IsMatcherOf}.
-     */
     @Test
-    public void containsErrorMessage() {
-        final Description msg = new StringDescription();
+    public void contains() {
```
Comment (llorllale) [inline src/test/java/org/llorllale/cactoos/matchers/HasValuesTest.java]: @dgroup I think this one can be called `matchSafely`

File: src/test/java/org/llorllale/cactoos/matchers/HasValuesTest.java [hunk 4]
```diff
@@ -42,81 +40,74 @@
  * @since 1.0.0
  * @checkstyle MagicNumberCheck (500 lines)
  * @checkstyle JavadocMethodCheck (500 lines)
```
Comment (llorllale) [inline src/test/java/org/llorllale/cactoos/matchers/HasValuesTest.java]: @dgroup please remove `JavadocMethodCheck` - I prefer having tests with javadocs

File: src/test/java/org/llorllale/cactoos/matchers/HasValuesTest.java [hunk 5]
```diff
@@ -42,81 +40,74 @@
...
+     * java.lang.AssertionError:
+     *   Expected: <[5]>
+     *   but: <[1, 2, 3]>
+     * }</pre>, where
+     *  - the "actual section" is "<[1, 2, 3]>"
+     *  - the "expected section" is "<[5]>".
+     */
     @Test
-    public void nonNull() {
+    public void thatActualSectionOfHamcrestResultMessageIsCorrect() {
```
Comment (llorllale) [inline src/test/java/org/llorllale/cactoos/matchers/HasValuesTest.java]: @dgroup I think this one can be called `describeActualValues`

File: src/main/java/org/llorllale/cactoos/matchers/HasValues.java (hunk 1 above)
Comment (dgroup) [reply src/main/java/org/llorllale/cactoos/matchers/HasValues.java]: @llorllale done.

File: src/main/java/org/llorllale/cactoos/matchers/HasValues.java (hunk 2 above)
Comment (dgroup) [reply src/main/java/org/llorllale/cactoos/matchers/HasValues.java]: @llorllale Done.

File: src/test/java/org/llorllale/cactoos/matchers/HasValuesTest.java (hunk 3 above)
Comment (dgroup) [reply src/test/java/org/llorllale/cactoos/matchers/HasValuesTest.java]: @llorllale done.

File: src/test/java/org/llorllale/cactoos/matchers/HasValuesTest.java (hunk 4 above)
Comment (dgroup) [reply src/test/java/org/llorllale/cactoos/matchers/HasValuesTest.java]: Done.

File: src/test/java/org/llorllale/cactoos/matchers/HasValuesTest.java (hunk 5 above)
Comment (dgroup) [reply src/test/java/org/llorllale/cactoos/matchers/HasValuesTest.java]: @llorllale done.

**Comment by llorllale** [comment]
@rultor merge

**Comment by 0crat** [comment]
Job `gh:llorllale/cactoos-matchers#43` is not assigned, can't get performer

**Comment by 0crat** [comment]
The job [#43](https://github.com/llorllale/cactoos-matchers/issues/43) is now out of scope

**Pull request merged**